ccm_web/client/src/api.ts

```diff
@@ -1,36 +1,63 @@
+import { CanvasCourseBase } from './models/canvas'
 import { Globals, HelloMessageData } from './models/models'
 import handleErrors from './utils/handleErrors'
 
 export interface LtiProps {
   ltiKey: string | undefined
 }
 
-const createAuthHeaders = (key: string | undefined): RequestInit => {
+const initRequest = (key: string | undefined, headers: string[][] = []): RequestInit => {
   if (key !== undefined) {
+    headers.push(['Authorization', 'Bearer ' + key])
     return {
       credentials: 'include',
-      headers: { Authorization: 'Bearer ' + key }
+      headers: headers
     }
   }
   return {}
 }
 
+const getGet = (key: string | undefined): RequestInit => {
+  const request = initRequest(key)
+  request.method = 'GET'
+  return request
+}
+
+// This currently assumes all put requests have a json payload and recieve a json response
+const getPut = (key: string | undefined, body: string): RequestInit => {
+  const headers: string[][] = []
+  headers.push(['Content-Type', 'application/json'])
+  headers.push(['Accept', 'application/json'])
+  const request = initRequest(key, headers)
+  request.method = 'PUT'
+  request.body = body
+  return request
+}
+
+export const getCourse = async (key: string | undefined, courseId: number): Promise<CanvasCourseBase> => {
+  const request = getGet(key)
+  const resp = await fetch(`/api/course/${courseId}/name`, request)
+  await handleErrors(resp)
+  return await resp.json()
```

Comment:
![Screen Shot 2021-07-01 at 1 10 27 PM](https://user-images.githubusercontent.com/8579775/124169455-42fefa00-da74-11eb-808f-c9b80a9eec5f.png)

I see this error when getting course name failed. should be leave it as blank if the call failed? or we can pull that info from LTI launch and give it.

I am OK to leave it blank if that's not available

Reply:
Another thing I noticed is that the `getCourseName` is called every time when user switches between features. To me that is redundant call when user navigating using breadcrumbs

I feel some how the course name should be saved in the React state or props once we get the course name and not make API call every time. since homepage is often visited while navigating between features so getCourseName call should not be made after success. UI should make another getCourseName call if it failed to get during the launch when Homepage is navigated using breadcrumbs

Reply:
Also whenever the PUT happens for updating the course name, the course name that is holding in the local react code should be updating as well.

Reply:
@pushyamig get course seems to take about .5 sec or less, testing locally for me.  Not insignificant but not huge.  Do you think putting course info into globals should be spun off into its own issue?

Reply:
What local react code are you referring to.. do you se an issue somewhere?  Not [this](https://github.com/chrisrrowland/canvas-course-manager-next/blob/fe8cd3a0395ea5036181043c385dca31623ede97/ccm_web/client/src/pages/Home.tsx#L61)?

Reply:
> What local react code are you referring to.. do you se an issue somewhere? Not this?

I mean to say when a course name get changes/updated then I feel it should be placed in the props. If SetCourse is the method, probably that is the place i guess. This suggestion is based on idea that courseName should be only pulled once during the launch of the tool and not when user navigates back to homepage via breadcrumbs

Reply:
> @pushyamig get course seems to take about .5 sec or less, testing locally for me. Not insignificant but not huge. Do you think putting course info into globals should be spun off into its own issue?

I want to hear from @ssciolla on this. I am ok it leave it off as part of this issue.

Reply:
Hm, some interesting questions raised in this discussion. I also agree that we probably want to avoid spamming the API every time we navigate back to `Home`. One way to solve that would be by moving the hook and initial API call up to `App` and then passing the `set` action down to the `Home` page component.

I don't know if I like putting the course object in `globals`, because if the course name changes, then we either need to re-fetch globals or change the `globals` object. I'd like to not modify or fetch globals a bunch. Keeping the course fetching separate seems better to me, but I could probably be persuaded if someone feels strongly.

Reply:
> One way to solve that would be by moving the hook and initial API call up to App and then passing the set action down to the Home page component.
I like  this idea!

> I don't know if I like putting the course object in globals, because if the course name changes, then we either need to re-fetch globals or change the globals object.

React [state-lifting](https://reactjs.org/docs/lifting-state-up.html) can be used with out refetching the globals

Reply:
I moved the getCourse call into App

Reply:
Thanks for making the change. In my head, I was also thinking that it might make sense to move the set API call up and pass that down, but you did exactly what I said and it works. I'm fine with what you did. However, let's watch out to see if it makes sense to move that other API action (`usePromise` hook for `setCourseName`) up in the future (it doesn't seem to yet).

Reply:
What is conclusion of what to display when fetching an getCourseName has an error? If we are waiting to here from @melindakraft and @janelilr on this, then we need to create an issue to deal with that later